Failing on TS 4.7 nightlies

There appears to be a change in how inference is working with `this` types on the latest 4.7 nightlies.

[Example run](https://github.com/true-myth/true-myth/runs/5698568682?check_suite_focus=true)

<details><summary>failing output</summary>

```
FAIL test/maybe.test.ts
9
  ● Test suite failed to run
10

11
    src/maybe.ts:244:21 - error TS2345: Argument of type 'Maybe<T>' is not assignable to parameter of type 'Maybe<{ toString(): string; }>'.
12
      Type 'Just<T>' is not assignable to type 'Maybe<{ toString(): string; }>'.
13
        Type 'Just<T>' is not assignable to type 'Just<{ toString(): string; }>'.
14
          Type 'T' is not assignable to type '{ toString(): string; }'.
15

16
    244     return toString(this);
17
                            ~~~~
18

19
      src/maybe.ts:53:17
20
        53 class MaybeImpl<T> {
21
                           ~
22
        This type parameter probably needs an `extends object` constraint.
23

24
FAIL test/result.test.ts
25
  ● Test suite failed to run
26

27
    src/maybe.ts:244:21 - error TS2345: Argument of type 'Maybe<T>' is not assignable to parameter of type 'Maybe<{ toString(): string; }>'.
28
      Type 'Just<T>' is not assignable to type 'Maybe<{ toString(): string; }>'.
29
        Type 'Just<T>' is not assignable to type 'Just<{ toString(): string; }>'.
30
          Type 'T' is not assignable to type '{ toString(): string; }'.
31

32
    244     return toString(this);
33
                            ~~~~
34

35
      src/maybe.ts:53:17
36
        53 class MaybeImpl<T> {
37
                           ~
38
        This type parameter probably needs an `extends object` constraint.
39

40
FAIL test/toolbelt.test.ts
41
  ● Test suite failed to run
42

43
    src/maybe.ts:244:21 - error TS2345: Argument of type 'Maybe<T>' is not assignable to parameter of type 'Maybe<{ toString(): string; }>'.
44
      Type 'Just<T>' is not assignable to type 'Maybe<{ toString(): string; }>'.
45
        Type 'Just<T>' is not assignable to type 'Just<{ toString(): string; }>'.
46
          Type 'T' is not assignable to type '{ toString(): string; }'.
47

48
    244     return toString(this);
49
                            ~~~~
50

51
      src/maybe.ts:53:17
52
        53 class MaybeImpl<T> {
53
                           ~
54
        This type parameter probably needs an `extends object` constraint.
55

56
PASS test/unit.test.ts
57
Failed to collect coverage from /home/runner/work/true-myth/true-myth/src/result.ts
58
ERROR: src/result.ts:216:21 - error TS2345: Argument of type 'Result<T, E>' is not assignable to parameter of type 'Result<{ toString(): string; }, { toString(): string; }>'.
59
  Type 'Ok<T, E>' is not assignable to type 'Result<{ toString(): string; }, { toString(): string; }>'.
60
    Type 'Ok<T, E>' is not assignable to type 'Ok<{ toString(): string; }, { toString(): string; }>'.
61
      Type 'T' is not assignable to type '{ toString(): string; }'.
62

63
216     return toString(this);
64
                        ~~~~
65

66
  src/result.ts:44:18
67
    44 class ResultImpl<T, E> {
68
                        ~
69
    This type parameter probably needs an `extends object` constraint.
70
STACK: 
71
Failed to collect coverage from /home/runner/work/true-myth/true-myth/src/maybe.ts
72
ERROR: src/maybe.ts:244:21 - error TS2345: Argument of type 'Maybe<T>' is not assignable to parameter of type 'Maybe<{ toString(): string; }>'.
73
  Type 'Just<T>' is not assignable to type 'Maybe<{ toString(): string; }>'.
74
    Type 'Just<T>' is not assignable to type 'Just<{ toString(): string; }>'.
75
      Type 'T' is not assignable to type '{ toString(): string; }'.
76

77
244     return toString(this);
78
                        ~~~~
79

80
  src/maybe.ts:53:17
81
    53 class MaybeImpl<T> {
82
                       ~
83
    This type parameter probably needs an `extends object` constraint.
```

</summary>

--- a/src/-private/utils.ts
+++ b/src/-private/utils.ts
@@ -16,3 +16,20 @@ export const isVoid = (value: unknown): value is undefined | null =>
 export function curry1<T, U>(op: (t: T) => U, item?: T) {
   return item !== undefined ? op(item) : op;
 }
+
+/**
+ * Check whether a given key is in an object
+ * @internal
+ */
+function has<T, K extends PropertyKey>(value: T, key: K): value is T & { [Key in K]: unknown } {
+  return typeof value === 'object' && value !== null && key in value;
+}
+
+export function safeToString(value: unknown): string {
+  if (has(value, 'toString') && typeof value['toString'] === 'function') {
+    const fnResult = value.toString();
+    return typeof fnResult === 'string' ? fnResult : JSON.stringify(value);
+  } else {
+    return JSON.stringify(value);
+  }
+}

--- a/src/maybe.ts
+++ b/src/maybe.ts
@@ -5,7 +5,7 @@
  */
 
 import Result from './result.js';
-import { curry1, isVoid } from './-private/utils.js';
+import { curry1, isVoid, safeToString } from './-private/utils.js';
 
 // Import for backwards-compatibility re-export
 import * as Toolbelt from './toolbelt.js';
@@ -846,8 +846,8 @@ export function fromResult<T>(result: Result<T, unknown>): Maybe<T> {
   @param maybe The value to convert to a string.
   @returns     The string representation of the `Maybe`.
  */
-export function toString<T extends { toString(): string }>(maybe: Maybe<T>): string {
-  const body = maybe.isJust ? `(${maybe.value.toString()})` : '';
+export function toString<T>(maybe: Maybe<T>): string {
+  const body = maybe.map((value) => `(${safeToString(value)})`).unwrapOr('');
   return `${maybe.variant}${body}`;
 }
 

--- a/src/result.ts
+++ b/src/result.ts
@@ -7,7 +7,7 @@
 import type Maybe from './maybe.js';
 
 import Unit from './unit.js';
-import { curry1, isVoid } from './-private/utils.js';
+import { curry1, isVoid, safeToString } from './-private/utils.js';
 
 // Import for backwards-compatibility re-export
 import * as Toolbelt from './toolbelt.js';
@@ -988,15 +988,13 @@ export function fromMaybe<T, E>(
   `toString(err([1, 2, 3]))`        | `Err(1,2,3)`
   `toString(err({ an: 'object' }))` | `Err([object Object])`
 
-  @typeparam T The type of the wrapped value; its own `.toString` will be used
-               to print the interior contents of the `Just` variant.
-  @param maybe The value to convert to a string.
-  @returns     The string representation of the `Maybe`.
+  @typeparam T  The type of the wrapped value; its own `.toString` will be used
+                to print the interior contents of the `Just` variant.
+  @param result The value to convert to a string.
+  @returns      The string representation of the `Maybe`.
  */
-export const toString = <T extends { toString(): string }, E extends { toString(): string }>(
-  result: Result<T, E>
-): string => {
-  const body = (result.isOk ? result.value : result.error).toString();
+export const toString = <T, E>(result: Result<T, E>): string => {
+  const body = result.match({ Ok: safeToString, Err: safeToString });
   return `${result.variant.toString()}(${body})`;
 };
 

--- a/test/maybe.test.ts
+++ b/test/maybe.test.ts
@@ -282,9 +282,33 @@ describe('`Maybe` pure functions', () => {
     expect(MaybeNS.fromResult(anErr)).toEqual(MaybeNS.nothing());
   });
 
-  test('`toString`', () => {
-    expect(MaybeNS.toString(MaybeNS.of(42))).toEqual('Just(42)');
-    expect(MaybeNS.toString(MaybeNS.nothing<string>())).toEqual('Nothing');
+  describe('`toString`', () => {
+    test('normal cases', () => {
+      expect(MaybeNS.toString(MaybeNS.of(42))).toEqual('Just(42)');
+      expect(MaybeNS.toString(MaybeNS.nothing<string>())).toEqual('Nothing');
+    });
+
+    test('custom `toString`s', () => {
+      const withNotAFunction = {
+        whyThough: 'because JS bro',
+        toString: '🤨',
+      };
+
+      expect(MaybeNS.toString(Maybe.of(withNotAFunction))).toEqual(
+        `Just(${JSON.stringify(withNotAFunction)})`
+      );
+
+      const withBadFunction = {
+        cueSobbing: true,
+        toString() {
+          return { lol: 123 };
+        },
+      };
+
+      expect(MaybeNS.toString(Maybe.of(withBadFunction))).toEqual(
+        `Just(${JSON.stringify(withBadFunction)})`
+      );
+    });
   });
 
   test('`toJSON`', () => {

--- a/test/result.test.ts
+++ b/test/result.test.ts
@@ -318,15 +318,39 @@ describe('`Result` pure functions', () => {
     expect(ResultNS.fromMaybe(errValue, aNothing)).toEqual(anErr);
   });
 
-  test('toString', () => {
-    const theValue = { thisIsReally: 'something' };
-    const errValue = ['oh', 'no'];
+  describe('toString', () => {
+    test('normal cases', () => {
+      const theValue = { thisIsReally: 'something' };
+      const errValue = ['oh', 'no'];
 
-    const anOk = ResultNS.ok<typeof theValue, typeof errValue>(theValue);
-    expect(ResultNS.toString(anOk)).toEqual(`Ok(${theValue.toString()})`);
+      const anOk = ResultNS.ok<typeof theValue, typeof errValue>(theValue);
+      expect(ResultNS.toString(anOk)).toEqual(`Ok(${theValue.toString()})`);
 
-    const anErr = ResultNS.err<typeof theValue, typeof errValue>(errValue);
-    expect(ResultNS.toString(anErr)).toEqual(`Err(${errValue.toString()})`);
+      const anErr = ResultNS.err<typeof theValue, typeof errValue>(errValue);
+      expect(ResultNS.toString(anErr)).toEqual(`Err(${errValue.toString()})`);
+    });
+
+    test('custom `toString`s', () => {
+      const withNotAFunction = {
+        whyThough: 'because JS bro',
+        toString: '🤨',
+      };
+
+      expect(ResultNS.toString(Result.ok(withNotAFunction))).toEqual(
+        `Ok(${JSON.stringify(withNotAFunction)})`
+      );
+
+      const withBadFunction = {
+        cueSobbing: true,
+        toString() {
+          return { lol: 123 };
+        },
+      };
+
+      expect(ResultNS.toString(Result.err(withBadFunction))).toEqual(
+        `Err(${JSON.stringify(withBadFunction)})`
+      );
+    });
   });
 
   test('`toJSON`', () => {
@@ -825,6 +849,6 @@ describe('`ResultNS.Err` class', () => {
 
     const result = fn.ap(val);
 
-    expect(result.toString()).toEqual(`Err(ERR_ALLURBASE)`);
+    expect(result.toString()).toEqual(`Err("ERR_ALLURBASE")`);
   });
 });

